libweb3jsonrpc/Test.cpp

```diff
@@ -23,16 +23,60 @@
 #include "Test.h"
 #include <jsonrpccpp/common/errors.h>
 #include <jsonrpccpp/common/exception.h>
-#include <libethereum/ClientTest.h>
+#include <libdevcore/CommonJS.h>
 #include <libethereum/ChainParams.h>
+#include <libethereum/ClientTest.h>
 
 using namespace std;
 using namespace dev;
+using namespace dev::eth;
 using namespace dev::rpc;
 using namespace jsonrpc;
 
 Test::Test(eth::Client& _eth): m_eth(_eth) {}
 
+namespace
+{
+string logEntriesToLogHash(eth::LogEntries const& _logs)
+{
+    RLPStream s;
+    s.appendList(_logs.size());
+    for (eth::LogEntry const& l : _logs)
+        l.streamRLP(s);
+    return toJS(sha3(s.out()));
+}
+}
+
+string Test::test_getLogHash(string const& _txHash)
+{
+    try
+    {
+        h256 txHash;
+        try
+        {
+            txHash = h256(_txHash);
+        }
+        catch (BadHexCharacter const&)
+        {
+            throw JsonRpcException(Errors::ERROR_RPC_INVALID_PARAMS);
+        }
+
+        if (m_eth.blockChain().isKnownTransaction(txHash))
+        {
+            LocalisedTransaction t = m_eth.localisedTransaction(txHash);
+            BlockReceipts const& blockReceipts = m_eth.blockChain().receipts(t.blockHash());
+            if (blockReceipts.receipts.size() != 0)
+                return logEntriesToLogHash(blockReceipts.receipts[t.transactionIndex()].log());
+        }
+        return toJS(dev::EmptyListSHA3);
+    }
+    catch (std::exception const& ex)
+    {
+        cwarn << ex.what();
```

Comment:
When something goes wrong we don't want to just silently return "internal error", we'd like to have a way to understand what happened, so let's output the error to the log.

Reply:
_ex ?

Reply:
you could still use boost exception for nicer error output like this: 
`BOOST_THROW_EXCEPTION(JsonRpcException(Errors::ERROR_RPC_INTERNAL_ERROR, _ex.what()));`

Reply:
No
1. `BOOST_THROW_EXCEPTION` wraps your exception with additional info like filename and line in debug build, but it can be useful only if it's caught and accessed with boost.exception facilities. Here `JsonRpcException` is always caught by our JSON-RPC framework, which doesn't know anything about boost.exception and just transforms it into JSON response.
2. The second argument to `JsonRpcException` constructor will also be added to JSON response, but `what()` string is usually for our exceptions not user-friendly at all, I don't think it will look appropriate at the client side. We should just log it in our log.

Reply:
`ex` is not a function parameter, so no `_` prefix needed